### PR TITLE
Add valid io driver

### PR DESCRIPTION
**Pull request by chick**
Adds Valid control structure similar to Decoupled
Adds tests of new construct

### CODE REVIEW

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
+package chisel3.tests
+
+import chisel3._
+import chisel3.tester._
+import chisel3.tester.experimental.TestOptionBuilder._
+
+import chisel3.util.Valid
+import org.scalatest._
+
+class ValidQueue[T <: Data](ioType: T) extends MultiIOModule {
```
[the remark this answers is not in the document]
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: Went with second option, just passed things straight through. Also simplified code

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  }.otherwise {
+    regData := regData
+  }
+
+  regValid := in.valid
+  out.bits := regData
+  out.valid := regValid
+}
+
+class ValidQueueModule extends MultiIOModule {
```
[the remark this answers is not in the document]
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: Uses Pipe now

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      c.in.initSource().setSourceClock(c.clock)
+      c.out.initSink().setSinkClock(c.clock)
+
+      c.out.expectInvalid()
+
+      fork {
+        c.in.enqueueSeq(Seq(42.U, 43.U, 44.U))
+      }.fork {
+        c.out.expectDequeue(42.U)
+        println("got 42")
```
[the remark this answers is not in the document]
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: removed

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        c.out.expectDequeue(43.U) // check that queue stalls
+        c.out.expectDequeue(44.U)
+        c.out.expectInvalid()
+      }.join()
+    }
+  }
+
+  it should "work with a combinational queue" in {
+    test(new ValidQueueModule) { c =>
+      c.in.initSource()
```
[the remark this answers is not in the document]
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: done

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        c.in.enqueueSeq(Seq(42.U, 43.U, 44.U))
+      }.fork {
+        c.out.expectDequeueSeq(Seq(42.U, 43.U, 44.U))
+      }.join()
+    }
+  }
+
+  it should "wait for data until valid data presented" in {
+    test(new ValidQueueModule) { c =>
+      c.in.initSource()
```
[the remark this answers is not in the document]
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: done

File: src/main/scala/chisel3/tester/ValidDriver.scala [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    ClockResolutionUtils.setClock(ValidDriver.decoupledSourceKey, x, clock)
+    this
+  }
+
+  protected def getSourceClock: Clock = {
+    ClockResolutionUtils.getClock(ValidDriver.decoupledSourceKey, x,
+      x.valid.getSourceClock)  // TODO: validate against bits/valid sink clocks
+  }
+
+  def enqueueNow(data: T): Unit = timescope {
```
[the remark this answers is not in the document]
Comment (chick) [reply src/main/scala/chisel3/tester/ValidDriver.scala]: I think this is take care of based on our conversation a couple of days ago.

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      c.in.initSource().setSourceClock(c.clock)
+      c.out.initSink().setSinkClock(c.clock)
+
+      c.out.expectInvalid()
+
+      fork {
+        c.in.enqueueSeq(Seq(42.U, 43.U, 44.U))
+      }.fork {
+        c.out.expectDequeue(42.U)
+        c.out.expectDequeue(43.U) // check that queue stalls
```
Comment (ducky64) [inline src/test/scala/chisel3/tests/ValidQueueTest.scala]: Wait does the queue stall?
Also why not use dequeueSeq?

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      c.out.initSink().setSinkClock(c.clock)
+
+      for(_ <- 0 until 10) {
+        c.out.expectInvalid()
+        c.clock.step()
+      }
+
+      c.in.enqueueNow(77.U)
+
+      c.out.expectInvalid()
```
Comment (ducky64) [inline src/test/scala/chisel3/tests/ValidQueueTest.scala]: Consider using a for loop to make behavior more parameterized?

File: src/main/scala/chisel3/tester/ValidDriver.scala [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    }.joinAndStep(getSinkClock)
+  }
+
+  def expectDequeueSeq(data: Seq[T]): Unit = timescope {
+    for (elt <- data) {
+      expectDequeue(elt)
+    }
+  }
+
+  def expectPeek(data: T): Unit = {
```
Comment (ducky64) [inline src/main/scala/chisel3/tester/ValidDriver.scala]: Maybe a comment here that emphasizes that this does not advance time, compared to expectDequeue

File: src/test/scala/chisel3/tests/ValidQueueTest.scala (hunk 7 above)
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: queue did not stall - that was a leftover comment
now uses expectDequeueSeq

File: .gitignore [hunk 10]
```diff
@@ -9,4 +9,4 @@
 .idea
 *.iml
 
-/test_run_dir/
+/test_run_dir_pre_1172/
```
Comment (ducky64) [inline .gitignore]: uh wut

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 11]
```diff
@@ -65,24 +63,49 @@ class ValidQueueTest extends FlatSpec with ChiselScalatestTester {
...
+      test(new ValidQueueModule(UInt(8.W), delay = delay)) { c =>
+        c.in.initSource().setSourceClock(c.clock)
+        c.out.initSink().setSinkClock(c.clock)
+
+        for (_ <- 0 until 10) {
+          c.out.expectInvalid()
+          c.clock.step()
+        }
+
+        for (value <- 70 to 80) {
```
Comment (ducky64) [inline src/test/scala/chisel3/tests/ValidQueueTest.scala]: This doesn't pipeline data, is that intentional?
If you want to pipeline data, consider rewriting using threads.

File: src/test/scala/chisel3/tests/ValidQueueTest.scala [hunk 12]
```diff
@@ -65,24 +63,49 @@ class ValidQueueTest extends FlatSpec with ChiselScalatestTester {
...
 
-      c.in.enqueueNow(77.U)
-
-      c.out.expectInvalid()
-      c.clock.step()
-      c.out.expectInvalid()
-      c.clock.step()
-
-      c.out.expectDequeueNow(77.U)
+  it should "expectInvalid can be used to wait for data" in {
```
Comment (ducky64) [inline src/test/scala/chisel3/tests/ValidQueueTest.scala]: wait how does this differ from the above test?

File: .gitignore (hunk 10 above)
Comment (chick) [reply .gitignore]: That's weird, it must have somehow caught a re-name i did on that dir to save some intermediate firrtl file. Undid it.

File: src/test/scala/chisel3/tests/ValidQueueTest.scala (hunk 12 above)
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: Looks like it got duplicated. Removed copied code.

File: src/test/scala/chisel3/tests/ValidQueueTest.scala (hunk 11 above)
Comment (chick) [reply src/test/scala/chisel3/tests/ValidQueueTest.scala]: Did not want to pipleline data. just wanted to show that expectInvalid works.

**Pull request merged**